pkg/admission/ingress/webhook_handler.go

```diff
@@ -0,0 +1,147 @@
+package dnsrecord
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+
+	// corev1 "k8s.io/api/core/v1"
+	// "net/http"
+
+	kuadrantv1 "github.com/Kuadrant/multi-cluster-traffic-controller/pkg/apis/v1"
+	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/dns"
+	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/traffic"
+	log "github.com/sirupsen/logrus"
+	networkingv1 "k8s.io/api/networking/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	controllertraffic "github.com/Kuadrant/multi-cluster-traffic-controller/pkg/controllers/traffic"
+)
+
+type Handler struct {
+	Decoder    *admission.Decoder
+	Serializer *json.Serializer
+	client.Client
+
+	Hosts        controllertraffic.HostService
+	Certificates controllertraffic.CertificateService
+	HostResolver dns.HostResolver
+}
+
+func CreateHandler(hostService controllertraffic.HostService, certService controllertraffic.CertificateService, hostResolver dns.HostResolver) (admission.Handler, error) {
+	log.Info("Creating handler ")
+	scheme := runtime.NewScheme()
+	if err := networkingv1.AddToScheme(scheme); err != nil {
+		log.Error("Unable add to scheme", err)
+		return nil, err
+	}
+
+	serialiser := json.NewSerializerWithOptions(json.DefaultMetaFactory, scheme, scheme, json.SerializerOptions{})
+
+	log.Info("Creating decoder")
+	decoder, err := admission.NewDecoder(scheme)
+	if err != nil {
+		log.Error("Unable to create decoder", err)
+		return nil, err
+	}
+	return &Handler{
+		Decoder:    decoder,
+		Serializer: serialiser,
+
+		Hosts:        hostService,
+		Certificates: certService,
+		HostResolver: hostResolver,
+	}, nil
+
+}
+
+func (h *Handler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log.Info("WEBHOOK HANDLER")
+
+	ingress := &networkingv1.Ingress{}
+	if err := h.Decoder.Decode(req, ingress); err != nil {
+		return admission.Errored(-1, err)
+	}
+
+	original := ingress.DeepCopy()
+
+	allowed, err := h.handle(ctx, ingress)
+	if err != nil {
+		admission.Errored(-1, err)
+	}
+
+	if !allowed {
+		return admission.Denied("")
+	}
+
+	if !equality.Semantic.DeepEqual(original, ingress) {
+		var originalSerialised bytes.Buffer
+		var currentSerialised bytes.Buffer
+
+		h.Serializer.Encode(original, &originalSerialised)
+		h.Serializer.Encode(ingress, &currentSerialised)
+
+		return admission.PatchResponseFromRaw(
+			originalSerialised.Bytes(),
+			currentSerialised.Bytes(),
+		)
+	}
+
+	return admission.Allowed("")
+
+}
+
+func (h *Handler) handle(ctx context.Context, ingress *networkingv1.Ingress) (bool, error) {
```

Comment:
This looks to be pulling the status from the Ingress to create entries in the DNS Record? I'm not sure the webhook is the best place for this? I think a lot of the logic in here should be instead run from the `pkg/controllers/traffic/traffic_controller.go`?

I have done a similar thing by calling out the traffic_controller logic from the syncer, in my syncer PR, see here: https://github.com/Kuadrant/multi-cluster-traffic-controller/pull/11/files#diff-cf3ce1c5bad0247b8bcf2c5d5eb789a53b4f752bc1684449472eca358dda2658R68

Reply:
The same too for setting the TLS certificate and maybe even the managed host? It all will still ultimately be a response from the webhook, but by moving the logic into the traffic_controller we should get the same logic when reconciling the ingress (i.e. not just for the webhook)

Reply:
My understanding is that we want to have that logic in both places (the reconciler _and_ the webhook). It is a bit redundant for now but I think the webhook will have more complex logic later on when we implement custom domain verification.

Maybe @maleck13 can confirm that

Reply:
So domain verification, host / dns record registration can be duplicated, but I would go for having the actual updating of the DNSRecord happen outside of the webhook if you follow?
Webhook can validate the ingress and create the initial DNSRecord and TLS certificte  to register a new host and mutate the ingress to add tls section

Reply:
my suggestion is rather: let's move the "duplicated" code to the traffic_controller, and call that from the webhook, effectively removing the duplication, anything that is webhook-specific should not be moved and instead stay here.

Reply:
or in other words:
```
webhook handle -> pass ingress to traffic_controller -> modify ingress with webhook specific logic -> return ingress on webhook response.
```

Reply:
traffic controller right now modifiies the ingress directly

Reply:
yes, I don't think that makes this incompatible? It doesn't send it to the API, just updates it in memory.

Reply:
I think there's still a few differences between the logic in the reconciler and the logic in the webhook. My understanding is that the webhook:
* Only creates (but never updates) the DNSRecords and Certificates for the Ingress
* Doesn't copy the TLS secret once it's created by cert-manager. It only sets the name in the TLS section of the Ingress

We could still probably refactor it to reduce the duplicated logic to a minimum but I'd leave that for another time I think

Reply:
ok, we can address this later.